### DISCUSSION TITLE
Firmware compilation help

### DISCUSSION THREAD

**Discussion by JDPKSP** [picocomputer/community#102] [Q&A] [answered]
Hi all.
I'd best start with acknowledging that this is very new to me. Nontheless I find myself wanting to compile the firmware myself (I need to add another keyboard layout).

I followed the dev setup and am building under Mint. (with vscode _not_ running in a container). It seems to compile fine, without errors, either by building in terminal or through vscode (F7). But the generatet .uf2s are a fraction of the expected size when compared to the release. And of course they do not function as intended when loaded to VGA and RIA.

I have not modified the source at all yet, so it's not that.
That being said, I suspect that, this being my first time attempting C, that I'm doing something obvious wrong.

In case it is any help, here is a dump of the build log:
https://gist.github.com/JDPKSP/33834fcbc0c1f5c65574ca71bafa3822

**Comment by vruumllc** [comment, marked as answer]
Here is what I do:

[cmake step] ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
tonyvr@tonyvr-laptop:~/projects/RP6502/rp6502$ rm -r build
rm: cannot remove 'build': No such file or directory
tonyvr@tonyvr-laptop:~/projects/RP6502/rp6502$ mkdir build
tonyvr@tonyvr-laptop:~/projects/RP6502/rp6502$ cd build
tonyvr@tonyvr-laptop:~/projects/RP6502/rp6502/build$ cmake ..
PICO_SDK_PATH is /home/tonyvr/projects/RP6502/rp6502/src/pico-sdk
Defaulting PICO_PLATFORM to rp2040 since not specified.
Defaulting PICO platform compiler to pico_arm_gcc since not specified.
-- Defaulting build type to 'Release' since not specified.
PICO compiler is pico_arm_gcc
-- The C compiler identification is GNU 10.3.1
-- The CXX compiler identification is GNU 10.3.1
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/arm-none-eabi-gcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/arm-none-eabi-g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- The ASM compiler identification is GNU
-- Found assembler: /usr/bin/arm-none-eabi-gcc
Build type is Release
Defaulting PICO target board to pico since not specified.
Using board configuration from /home/tonyvr/projects/RP6502/rp6502/src/pico-sdk/src/boards/include/boards/pico.h
-- Found Python3: /usr/bin/python3.10 (found version "3.10.12") found components: Interpreter 
TinyUSB available at /home/tonyvr/projects/RP6502/rp6502/src/pico-sdk/lib/tinyusb/src/portable/raspberrypi/rp2040; enabling build support for USB.
BTstack available at /home/tonyvr/projects/RP6502/rp6502/src/pico-sdk/lib/btstack
cyw43-driver available at /mnt/data/Projects/RP6502/rp6502/src/pico-sdk/lib/cyw43-driver
Pico W Bluetooth build support available.
lwIP available at /home/tonyvr/projects/RP6502/rp6502/src/pico-sdk/lib/lwip
mbedtls available at /home/tonyvr/projects/RP6502/rp6502/src/pico-sdk/lib/mbedtls
-- Configuring done
-- Generating done
-- Build files have been written to: /home/tonyvr/projects/RP6502/rp6502/build

[make step] ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
tonyvr@tonyvr-laptop:~/projects/RP6502/rp6502/build$ make > build.log
/home/tonyvr/projects/RP6502/rp6502/src/littlefs/lfs.c:495:13: warning: 'lfs_mlist_isopen' defined but not used [-Wunused-function]
  495 | static bool lfs_mlist_isopen(struct lfs_mlist *head

Here is the build.log generated:
[build.log](https://github.com/picocomputer/community/files/15046042/build.log)

[results] //////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
The resulting .uf2 files are found in build/src/, and are 360.4 k for rp6402_ria.uf2 and 163.8 k for rp6502_vga.uf2

These are also much smaller than the release .7 sizes, but work fine for me. I have no idea why the ones in the release zip file are so huge.

Some thoughts (without spending much time analyzing your log)
1. You are doing a debug build. I haven't tried this in a while. Try Release.
2. Did you do the submodule updates after the git clone?

git submodule update --init
cd src/pico-sdk
git submodule update --init
cd ../..

**Reply by JDPKSP** [reply]
Thank you for your reply.
It turns out I did indeed commit a somewhat silly error. The major differences in file sizes led to a somewhat lessened confidence with the generated files. So when I didn't see a COM port when connecting the board I concluded that they must not work.
I should have just waited a couple of seconds, then the COM port would have showed up.
My .uf2's are identical in size to yours, and they do indeed work. They do however seem to result in a somewhat slower startup time.

Could it be that @rumbledethumps uses some build optimizations that we do not? If so, it would be a great addition to the dev setup section.

And regarding the rest.
1. Debug was an oversight on my part. I've now set Cmake tools to default to Release
2. I did at first suspect submodule issues, so apart from following the guide, i ultimately performed a recursive update. No change btw.

**Reply by vruumllc** [reply]
Glad you are up and running.

**Comment by JDPKSP** [comment]
I figured the size discrepancy out. Turns out all other codepages than 850 are omitted by default.
in:
src/CMakeLists.txt
set:
RP6502_CODE_PAGE=0
to include all codepages
